src/data/option/n_ary.lean

```diff
@@ -0,0 +1,153 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.option.basic
+
+/-!
+# Binary map of options
+
+This file defines the binary map of `option`. This is mostly useful to define pointwise operations
+on intervals.
+
+## Main declarations
+
+* `option.map₂`: Binary map of options.
+
+## Notes
+
+This file is very similar to the n-ary section of `data.set.basic`, to `data.finset.n_ary` and to
+`order.filter.n_ary`. Please keep them in sync.
+
+We do not define `option.map₃` as its only purpose would be to prove properties of `option.map₂`
+and casing already fulfills this task.
+-/
+
+namespace option
+variables {α α' β β' γ γ' δ δ' ε ε' : Type*} {f : α → β → γ} {a : option α} {b : option β}
+  {c : option γ}
+
+/-- The image of a binary function `f : α → β → γ` as a function `option α → option β → option γ`.
+Mathematically this should be thought of as the image of the corresponding function `α × β → γ`. -/
+def map₂ (f : α → β → γ) (a : option α) (b : option β) : option γ := a.bind $ λ a, b.map $ f a
```

Comment:
It would be nice to show this definition is equal to `f <$> a <*> b`, which is the more general version not specialised to `option`.

Reply:
Note that it's not actually more general due to the lack of universe polymorphism.

Reply:
I'm well aware - that's precisely why I didn't say you should change the definition.